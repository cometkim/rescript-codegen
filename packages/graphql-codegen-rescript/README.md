# graphql-codegen-rescript

A [GraphQL Code Generator](https://the-guild.dev/graphql/codegen) plugin to generate common types for [ReScript](https://rescript-lang.org)

## LICENSE

MIT
