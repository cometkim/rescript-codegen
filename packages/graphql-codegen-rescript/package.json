{
  "name": "graphql-codegen-rescript",
  "version": "0.0.1",
  "license": "MIT",
  "keywords": [
    "codegen",
    "graphql",
    "graphql-codegen",
    "rescript",
    "tooling"
  ],
  "exports": {
    ".": "./src/index.ts"
  },
  "publishConfig": {
    "exports": {
      ".": {
        "types": "./dist/index.d.ts",
        "require": "./dist/index.cjs",
        "import": "./dist/index.mjs"
      },
      "./package.json": "./package.json"
    }
  },
  "scripts": {
    "test": "vitest",
    "build": "nanobundle build"
  },
  "files": [
    "src",
    "dist"
  ],
  "engines": {
    "node": ">=18"
  },
  "peerDependencies": {
    "graphql": "^15.0.0 || ^16.0.0"
  },
  "dependencies": {
    "@graphql-codegen/plugin-helpers": "^5.0.1",
    "@graphql-codegen/visitor-plugin-common": "^4.0.1"
  },
  "devDependencies": {
    "@graphql-codegen/core": "^4.0.0",
    "graphql": "^16.8.1",
    "nanobundle": "^2.0.0",
    "typescript": "^5.3.3",
    "vitest": "^1.1.1"
  }
}
