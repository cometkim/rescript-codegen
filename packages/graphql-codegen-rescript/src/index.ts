import { visit } from 'graphql';
import { indentMultiline } from '@graphql-codegen/visitor-plugin-common';
import { getCachedDocumentNodeFromSchema, type PluginFunction } from '@graphql-codegen/plugin-helpers';

import { type PluginConfig } from './config.ts';
import { ReScriptTypesVisitor } from './visitor.ts';

export { PluginConfig };

export const plugin: PluginFunction<PluginConfig> = (schema, _documents) => {
  const visitor = new ReScriptTypesVisitor(schema, {}, {});

  const ast = getCachedDocumentNodeFromSchema(schema);
  const result = visit(ast, visitor);
  const fragments = result.definitions
    .filter((d: any) => typeof d === 'string') as unknown as string[];

  const generated = fragments.join('\n\n');

  return {
    prepend: [
      'module Types = {',
      indentMultiline(
        visitor.getCommonDefinitions().join('\n'),
      ),
    ],
    content: indentMultiline(generated),
    append: ['}'],
  };
};
