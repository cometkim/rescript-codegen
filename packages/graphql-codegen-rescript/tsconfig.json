{
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "composite": true,
    "emitDeclarationOnly": true,
    "allowImportingTsExtensions": true,
    "strict": true,
    "outDir": "dist"
  }
}
