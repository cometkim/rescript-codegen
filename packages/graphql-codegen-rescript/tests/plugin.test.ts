import { describe, test, expect } from 'vitest';
import { buildSchema } from 'graphql';
import { codegen } from '@graphql-codegen/core';

import * as plugin from 'graphql-codegen-rescript';

const gql = String.raw;

describe('graphql-codegen-rescript', () => {
  async function executePlugin(source: string, config?: plugin.PluginConfig) {
    const schemaAst = buildSchema(source);
    const result = await codegen({
      schema: null as any,
      schemaAst,
      filename: '',
      documents: [],
      plugins: [
        {
          rescript: config ?? {},
        },
      ],
      pluginMap: {
        rescript: plugin,
      },
      config: {
      },
    });
    return result;
  }

  test('success', async () => {
    const result = await executePlugin(gql`
      scalar Date

      schema {
        query: Query
      }

      type Query {
        me: User!
        user(id: ID!): User
        allUsers: [User]
        search(term: String!): [SearchResult!]!
        myChats: [Chat!]!
      }

      enum Role {
        USER
        ADMIN
      }

      interface Node {
        id: ID!
      }

      union SearchResult = User | Chat | ChatMessage

      type User implements Node {
        id: ID!
        username: String!
        email: String!
        role: Role!
      }

      type Chat implements Node {
        id: ID!
        users: [User!]!
        messages: [ChatMessage!]!
      }

      type ChatMessage implements Node {
        id: ID!
        content: String!
        time: Date!
        user: User!
      }
    `);

    expect(result).toMatchSnapshot();
  });
});
