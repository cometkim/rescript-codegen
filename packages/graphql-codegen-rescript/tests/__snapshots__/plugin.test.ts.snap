// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graphql-codegen-rescript > success 1`] = `
"module Types = {
  type unknown
  
  type rec \\"Date" = unknown
  
  and \\"Query.me" = \\"User"
  and \\"Query.user" = option<\\"User">
  and \\"Query.allUsers" = option<array<option<\\"User">>>
  and \\"Query.search" = array<\\"SearchResult">
  and \\"Query.myChats" = array<\\"Chat">
  and \\"Query" = {
    me: \\"Query.me",
    user: \\"Query.user",
    allUsers: \\"Query.allUsers",
    search: \\"Query.search",
    myChats: \\"Query.myChats",
  }
  
  and \\"Role" =
    | USER
    | ADMIN
  
  and \\"Node.id" = id
  and \\"Node" = {
    id: \\"Node.id",
  }
  
  @tag("__typename")
  and \\"SearchResult" =
    | User(\\"User")
    | Chat(\\"Chat")
    | ChatMessage(\\"ChatMessage")
  
  and \\"User.id" = id
  and \\"User.username" = string
  and \\"User.email" = string
  and \\"User.role" = \\"Role"
  and \\"User" = {
    id: \\"User.id",
    username: \\"User.username",
    email: \\"User.email",
    role: \\"User.role",
  }
  
  and \\"Chat.id" = id
  and \\"Chat.users" = array<\\"User">
  and \\"Chat.messages" = array<\\"ChatMessage">
  and \\"Chat" = {
    id: \\"Chat.id",
    users: \\"Chat.users",
    messages: \\"Chat.messages",
  }
  
  and \\"ChatMessage.id" = id
  and \\"ChatMessage.content" = string
  and \\"ChatMessage.time" = \\"Date"
  and \\"ChatMessage.user" = \\"User"
  and \\"ChatMessage" = {
    id: \\"ChatMessage.id",
    content: \\"ChatMessage.content",
    time: \\"ChatMessage.time",
    user: \\"ChatMessage.user",
  }
}"
`;
